File: scripts/session/cleanup.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'time'
require 'optparse'

module Gmux
  module Scripts
    class SessionCleanup
      STORE_PATH = File.join(Dir.home, '.gmux', 'sessions.json')
      DEFAULT_MAX_AGE_DAYS = 7

      def initialize(options = {})
        @max_age_days = options[:max_age_days] || DEFAULT_MAX_AGE_DAYS
        @dry_run = options[:dry_run] || false
        @verbose = options[:verbose] || false
        @force = options[:force] || false
      end

      def run
        puts "=== GMUX Session Cleanup ==="
        puts "Max age: #{@max_age_days} days"
        puts "Dry run: #{@dry_run}"
        puts

        sessions = load_sessions
        if sessions.empty?
          puts "No sessions found."
          return
        end

        puts "Found #{sessions.length} session(s)"
        puts

        stale_sessions = find_stale_sessions(sessions)
        if stale_sessions.empty?
          puts "No stale sessions found."
          return
        end

        puts "Found #{stale_sessions.length} stale session(s):"
        stale_sessions.each do |name, session|
          age = session_age(session)
          puts "  - #{name} (#{age.round(1)} days old, status: #{session['status']})"
        end
        puts

        if @dry_run
          puts "[DRY RUN] Would remove #{stale_sessions.length} session(s)"
          return
        end

        unless @force
          print "Remove #{stale_sessions.length} stale session(s)? [y/N] "
          return unless STDIN.gets.chomp.downcase == 'y'
        end

        remove_sessions(stale_sessions)
        puts "Done. Removed #{stale_sessions.length} session(s)."
      end

      private

      def load_sessions
        return {} unless File.exist?(STORE_PATH)

        JSON.parse(File.read(STORE_PATH))
      rescue JSON::ParserError => e
        puts "Error parsing sessions file: #{e.message}"
        {}
      end

      def find_stale_sessions(sessions)
        now = Time.now
        sessions.select do |name, session|
          started_at = Time.parse(session['startedAt'])
          age_days = (now - started_at) / (24 * 3600)
          age_days > @max_age_days
        end
      end

      def session_age(session)
        now = Time.now
        started_at = Time.parse(session['startedAt'])
        (now - started_at) / (24 * 3600)
      end

      def remove_sessions(stale_sessions)
        stale_sessions.each do |name, session|
          remove_worktree(session['worktreePath']) if session['worktreePath']
          remove_tmux_window(session['tmuxWindowId']) if session['tmuxWindowId']
          remove_session_record(name)
          puts "  Removed: #{name}" if @verbose
        end
      end

      def remove_worktree(path)
        return unless File.directory?(path)

        puts "  Removing worktree: #{path}" if @verbose
        FileUtils.rm_rf(path)
      rescue StandardError => e
        puts "  Warning: Failed to remove worktree #{path}: #{e.message}"
      end

      def remove_tmux_window(window_id)
        puts "  Killing tmux window: #{window_id}" if @verbose
        system("tmux kill-window -t #{window_id} 2>/dev/null")
      end

      def remove_session_record(name)
        sessions = load_sessions
        sessions.delete(name)
        File.write(STORE_PATH, JSON.pretty_generate(sessions))
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: #{$PROGRAM_NAME} [options]"

    opts.on('-d', '--max-days DAYS', Integer, "Max age in days (default: 7)") do |d|
      options[:max_age_days] = d
    end

    opts.on('-n', '--dry-run', "Show what would be removed without doing it") do
      options[:dry_run] = true
    end

    opts.on('-v', '--verbose', "Show detailed output") do
      options[:verbose] = true
    end

    opts.on('-f', '--force', "Skip confirmation prompt") do
      options[:force] = true
    end

    opts.on('-h', '--help', "Show this help") do
      puts opts
      exit
    end
  end.parse!

  Gmux::Scripts::SessionCleanup.new(options).run
end

File: scripts/session/cleanup.py
#!/usr/bin/env python3
"""GMUX Session Cleanup - Remove stale sessions/worktrees older than N days."""

import json
import logging
import os
import shutil
import subprocess
import sys
import time
from argparse import ArgumentParser
from datetime import datetime, timezone
from pathlib import Path

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

STORE_PATH = Path.home() / ".gmux" / "sessions.json"
DEFAULT_MAX_AGE_DAYS = 7


class SessionCleanup:
    def __init__(self, options: dict):
        self.max_age_days = options.get("max_age_days", DEFAULT_MAX_AGE_DAYS)
        self.dry_run = options.get("dry_run", False)
        self.verbose = options.get("verbose", False)
        self.force = options.get("force", False)

    def run(self):
        print("=== GMUX Session Cleanup ===")
        print(f"Max age: {self.max_age_days} days")
        print(f"Dry run: {self.dry_run}")
        print()

        sessions = self._load_sessions()
        if not sessions:
            print("No sessions found.")
            return

        print(f"Found {len(sessions)} session(s)")
        print()

        stale = self._find_stale_sessions(sessions)
        if not stale:
            print("No stale sessions found.")
            return

        print(f"Found {len(stale)} stale session(s):")
        for name, session in stale:
            age = self._session_age(session)
            print(f"  - {name} ({age:.1f} days old, status: {session.get('status')})")
        print()

        if self.dry_run:
            print(f"[DRY RUN] Would remove {len(stale)} session(s)")
            return

        if not self.force:
            response = input(f"Remove {len(stale)} stale session(s)? [y/N] ")
            if response.strip().lower() != "y":
                print("Aborted.")
                return

        self._remove_sessions(stale)
        print(f"Done. Removed {len(stale)} session(s).")

    def _load_sessions(self) -> dict:
        if not STORE_PATH.exists():
            return {}
        try:
            return json.loads(STORE_PATH.read_text())
        except (json.JSONDecodeError, OSError) as e:
            print(f"Error parsing sessions file: {e}")
            return {}

    def _find_stale_sessions(self, sessions: dict) -> list:
        now = datetime.now(timezone.utc)
        stale = []
        for name, session in sessions.items():
            started_str = session.get("startedAt")
            if not started_str:
                continue
            try:
                started = datetime.fromisoformat(started_str)
                age_days = (now - started).total_seconds() / (24 * 3600)
                if age_days > self.max_age_days:
                    stale.append((name, session))
            except (ValueError, TypeError):
                continue
        return stale

    def _session_age(self, session: dict) -> float:
        started_str = session.get("startedAt")
        if not started_str:
            return 0.0
        try:
            started = datetime.fromisoformat(started_str)
            return (datetime.now(timezone.utc) - started).total_seconds() / (24 * 3600)
        except (ValueError, TypeError):
            return 0.0

    def _remove_sessions(self, stale: list):
        for name, session in stale:
            self._remove_worktree(session.get("worktreePath"))
            self._remove_tmux_window(session.get("tmuxWindowId"))
            self._remove_session_record(name)
            if self.verbose:
                print(f"  Removed: {name}")

    def _remove_worktree(self, path):
        if not path or not os.path.isdir(path):
            return
        if self.verbose:
            print(f"  Removing worktree: {path}")
        shutil.rmtree(path, ignore_errors=True)

    def _remove_tmux_window(self, window_id):
        if not window_id:
            return
        if self.verbose:
            print(f"  Killing tmux window: {window_id}")
        subprocess.run(
            ["tmux", "kill-window", "-t", window_id],
            capture_output=True,
        )

    def _remove_session_record(self, name: str):
        sessions = self._load_sessions()
        sessions.pop(name, None)
        STORE_PATH.write_text(json.dumps(sessions, indent=2))


def main():
    parser = ArgumentParser(description="Remove stale GMUX sessions")
    parser.add_argument(
        "-d", "--max-days",
        type=int,
        default=DEFAULT_MAX_AGE_DAYS,
        help=f"Max age in days (default: {DEFAULT_MAX_AGE_DAYS})",
    )
    parser.add_argument(
        "-n", "--dry-run",
        action="store_true",
        help="Show what would be removed without doing it",
    )
    parser.add_argument(
        "-v", "--verbose",
        action="store_true",
        help="Show detailed output",
    )
    parser.add_argument(
        "-f", "--force",
        action="store_true",
        help="Skip confirmation prompt",
    )

    args = parser.parse_args()
    SessionCleanup(vars(args)).run()


if __name__ == "__main__":
    main()
